### PR TITLE
Feature/198 text field component

### PR DESCRIPTION
**Pull request by anna-lach**
Styled `sl-input` component
Styled `sl-label` component
Styled hint and validation

Contains changes after tokens grouping.

### CODE REVIEW

File: packages/components/input/src/input.scss [hunk 1]
```diff
@@ -1,58 +1,231 @@
...
+      --_padding: var(--sl-space-input-text-field-block-#{$size})
+        var(--sl-space-input-text-field-inline-#{$size})
+        var(--sl-space-input-text-field-block-#{$size})
+        var(--sl-space-input-text-field-inline-#{$size});
+      --_gap: var(--sl-space-input-text-field-gap-#{$size});
+    }
+  }
+}
+
+@each $variant in $variants {
```
Comment (jpzwarte) [inline packages/components/input/src/input.scss:99]: So in other stylesheets, I've moved this block of code to the top; between `:host {` and the actual component styling. So all the CSS variable stuff is grouped together.

File: packages/components/input/src/input.stories.ts [hunk 2]
```diff
@@ -18,21 +18,62 @@ export const API: StoryObj = {
...
+      .size=${size}
+      .min=${min}
+      .max=${max}
+      .step=${step}
       .maxLength=${maxLength}
       .minLength=${minLength}
       .placeholder=${placeholder}
       .value=${value}
+      .type=${type}
+      ?readonly=${readonly}
```
Comment (jpzwarte) [inline packages/components/input/src/input.stories.ts:76]: Can you move this to the top with the other boolean properties?

File: packages/components/input/src/input.stories.ts [hunk 3]
```diff
@@ -41,7 +82,78 @@ export const API: StoryObj = {
...
 
 export const Disabled: StoryObj = {
-  render: () => html`<sl-input disabled value="I am disabled"></sl-input>`
+  render: () => html`
+    <style>
+      sl-input {
+        width: 400px;
+      }
+      </style>
+    </style><sl-input disabled value="I am disabled"></sl-input>`
```
Comment (jpzwarte) [inline packages/components/input/src/input.stories.ts:91]: Typo I assume :)

File: packages/components/input/src/input.ts [hunk 4]
```diff
@@ -147,14 +249,28 @@ export class Input extends FormControlMixin(HintMixin(LitElement)) {
     return html`
       <div @input=${this.#onInput} class="wrapper">
         <slot name="prefix"></slot>
-        <slot @slotchange=${this.#onSlotchange} name="input"></slot>
-        <slot name="suffix"></slot>
+        <slot
+          @slotchange=${this.#onSlotchange}
+          name="input"
+          @keydown=${this.#onKeydown}
```
Comment (jpzwarte) [inline packages/components/input/src/input.ts:255]: Please sort the event handlers alphabetically

File: packages/components/input/src/input.ts [hunk 5]
```diff
@@ -53,24 +74,54 @@ export class Input extends FormControlMixin(HintMixin(LitElement)) {
...
+  @property() type: 'email' | 'number' | 'tel' | 'text' | 'url' = 'text'; // TODO: password type will be added in the future
 
   /** Custom validators specified by the user. */
   @property({ attribute: false }) validators?: Validator[];
 
   /** The value for the input. */
   @property() value?: string;
 
+  /** @private used for focusVisible */
+  #clicked = false;
```
Comment (jpzwarte) [inline packages/components/input/src/input.ts:123]: Can you explain this bit?

File: packages/components/radio-group/src/radio-group.stories.ts [hunk 6]
```diff
@@ -237,7 +237,9 @@ export const CustomValidation: StoryObj = {
         <sl-radio value="1">One</sl-radio>
         <sl-radio value="2">Two</sl-radio>
         <sl-radio value="3">Three</sl-radio>
-        <div slot="value-missing">This is the custom value-missing message (for the required attribute).</div>
+        <div slot="value-missing" error-size="sm">
```
Comment (jpzwarte) [inline packages/components/radio-group/src/radio-group.stories.ts:240]: Should the error size be configurable? It feels like too much?

File: packages/components/shared/src/controllers/validation.ts [hunk 7]
```diff
@@ -32,9 +32,29 @@ export type ValidationConfig = {
...
+
 export const validationStyles: CSSResultGroup = css`
-  slot[part='error'] {
-    color: #c00;
+  slot[part='error']::slotted(*) {
+    color: var(--sl-color-input-helper-text-invalid);
+    display: inline-flex;
+    align-items: center;
+    fill: var(--sl-color-input-helper-icon-invalid);
+  }
```
Comment (jpzwarte) [inline packages/components/shared/src/controllers/validation.ts:43]: Please add newlines between rule blocks

**Review by jpzwarte — APPROVED** [review]

File: packages/components/input/src/input.ts (hunk 5 above)
Comment (jpzwarte) [reply packages/components/input/src/input.ts:123]: So affairs this goes against the browser heuristics. For example see the demo here: https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible

I really don't like going *against* browser behavior. 😕

File: packages/components/radio-group/src/radio-group.stories.ts (hunk 6 above)
Comment (jpzwarte) [reply packages/components/radio-group/src/radio-group.stories.ts:240]: Feels like the wrong approach for something that is pure styling. What about using a custom property?

```css
sl-input.mine {
  --sl-error-size: sm;
}
```

But you need container queries (style specifically) for that to work.

Perhaps
```
sl-input.mine {
  --sl-error-size: var(--sl-error-size-sm);
}
``` ?

**Review by Diaan — CHANGES_REQUESTED** [review]

File: packages/components/input/src/input.scss [hunk 8]
```diff
@@ -1,58 +1,231 @@
...
+
+      /* stylelint-disable-next-line max-nesting-depth */
+      &::placeholder {
+        --_placeholder-color: var(--sl-color-input-text-field-#{$variant}-default-placeholder);
+      }
+    }
+
+    slot[name='suffix'] {
+      /* stylelint-disable-next-line max-nesting-depth */
+      sl-icon {
```
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/input/src/input.scss:47]: right :)

File: packages/components/radio-group/src/radio-group.stories.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/radio-group/src/radio-group.stories.ts:240]: We can use the default one, it was just an example :)

File: packages/components/shared/src/controllers/validation.ts [hunk 9]
```diff
@@ -125,13 +163,20 @@ export class ValidationController implements ReactiveController {
     this.#host = host;
     this.#host.addController(this);
 
+    if (this.#host.hasAttribute('error-size')) {
+      this.#messageSize = this.#host.getAttribute('error-size') as MessageSize;
+    } else {
+      this.#messageSize = 'md';
+    }
```
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/shared/src/controllers/validation.ts:173]: Prettier handled it :D

File: packages/components/label/src/label.ts [hunk 10]
```diff
@@ -90,8 +105,10 @@ export class Label extends LitElement {
     return html`
       <slot @slotchange=${this.#onSlotchange} style="display: none"></slot>
       <slot name="label"></slot>
+      <slot name="icon"></slot>
+      <slot name="tooltip"></slot>
       ${this.optional ? html`<span class="optional">(optional)</span>` : ''}
-      ${this.required ? html`<span class="required">*</span>` : ''}
+      ${this.required ? html`<span class="required">(required)</span>` : ''}
```
[the remark this answers is not in the document]
Comment (anna-lach) [reply packages/components/label/src/label.ts:111]: good point, I need to add it here as well

**Review by jpzwarte — APPROVED** [review]

**Review by Diaan — APPROVED** [review]

**Pull request merged**